src/hooks/useFetch.js

```diff
@@ -4,103 +4,91 @@
  */
 
 import React from 'react';
-import { Client as BugsnagClient } from 'bugsnag-react-native';
 
-import { getAuthHeader, AUTH_ERROR_CODES } from 'sussol-utilities';
-
-import { SETTINGS_KEYS } from '../settings';
-import { UIDatabase } from '../database';
-
-const { SYNC_SITE_NAME, SYNC_SITE_PASSWORD_HASH } = SETTINGS_KEYS;
-
-const CONNECTION_TIMEOUT_PERIOD = 10000;
-
-const ERROR_CODES = {
-  ...AUTH_ERROR_CODES,
-  EMPTY_RESPONSE: 'No records found',
-};
-
-const bugsnagClient = new BugsnagClient();
-
-class BugsnagError extends Error {
-  constructor(message, data, ...args) {
-    super(message, ...args);
-    bugsnagClient.notify(this, report => {
-      report.errorMessage = message;
-      report.metadata = data;
-    });
-  }
-}
-
-const getAuthorizationHeader = () => {
-  const username = UIDatabase.getSetting(SYNC_SITE_NAME);
-  const password = UIDatabase.getSetting(SYNC_SITE_PASSWORD_HASH);
-  return getAuthHeader(username, password);
-};
-
-const isValidError = error => Object.values(ERROR_CODES).includes(error?.message);
+const doIf = (ifThis, doThis, ...withThese) => (ifThis() ? doThis(...withThese) : null);
 
 /**
  * Custom hook to fetch data inside mounted component.
  *
- * Fetches data on component mount. If the calling component is unmounted or the timeout is reached,
+ * If the calling component is unmounted or the timeout is reached,
  * the fetch is aborted.
  */
-export const useFetch = (url, options = {}, timeout = CONNECTION_TIMEOUT_PERIOD) => {
-  const [data, setData] = React.useState(null);
+export const useFetch = url => {
+  const _timer = React.useRef(null);
+  const _controller = React.useRef(null);
+
+  const _isMounted = React.useRef(false);
+  const _isBlocked = React.useRef(false);
+
+  const [isLoading, setIsLoading] = React.useState(false);
+  const [response, setResponse] = React.useState(null);
   const [error, setError] = React.useState(null);
-  const isMounted = React.useRef(true);
 
-  /* eslint-disable no-undef */
-  const abortController = new AbortController();
+  const reset = () => {
```

Comment:
This should be clearer now. Idea for `isBlocked` was to basically prevent race conditions or multiple fetch calls from causing errors/unexpected behaviour. On revisiting the code, I think I went a bit OTT with this, to the extent that it was duplicating some stuff that was already handled by `isMounted` and became quite confusing and probably risked introducing errors as much as preventing them. 

After refactor, `isBlocked` has been reduced to a simple semaphore to prevent multiple fetches being dispatched at once. It has a similar function to `isLoading`, but at the logic rather than UI level (can think of it sort of like getting a lock on fetching data). Main reason for it being a ref is to make it independent from the render cycle, just in case of race conditions or anything like that.

Reply:
nit: it's now `resetState` 🤣